docs/layout.rst

```diff
@@ -0,0 +1,155 @@
+.. _layout:
+
+Tensor layout
+=============
+
+Tensors (:ref:`scalar_tensor`) can be *placed* in a specific shape and *layout*.
+Having a good layout can be the key to performance.
+
+In Taichi, placing a layout is treated in a recursive manner.
+
+For example, this declares a 0-D tensor:
+
+.. code-block:: python
+
+    x = ti.var(ti.f32)
+    ti.root.place(x)
+    # or
+    x = ti.var(ti.f32, shape=())
+
+This declares a 1-D tensor of size ``3``:
+
+.. code-block:: python
+
+    x = ti.var(ti.f32)
+    ti.root.dense(ti.i, 3).place(x)
+    # or
+    x = ti.var(ti.f32, shape=3)
+
+This declares a 1-D tensor of shape ``(3, 4)``:
+
+.. code-block:: python
+
+    x = ti.var(ti.f32)
+    ti.root.dense(ti.ij, (3, 4)).place(x)
+    # or
+    x = ti.var(ti.f32, shape=(3, 4))
+
+Now, you may say, why not just simply specify ``shape=`` argument to specify tensor shape? Why we need these fancy stuffs like ``ti.root`` and ``place``? Wasn't them the same?
+Good question, let go forward and figure out why.
+
+
+Multi-dimentional layout
+------------------------
+
+As we all know, most mordern computers have a 1-D memory model.
+So it's not a big deal to store 1-D tensors, the address of ``i``-th can simply be ``base + i``.
+
+To store a multi-D tensor, however, it has to be reshaped into 1-D, to fit into the 1-D address space.
+For example, to store a 2-D tensor of size ``(3, 2)``, there are two way to do this:
+
+    1. The address of ``(i, j)``-th is ``base + i * 2 + j``.
+    1. The address of ``(i, j)``-th is ``base + j * 3 + i``.
+
+To specify which layout to use in Taichi:
+
+.. code-block:: python
+
+    ti.root.dense(ti.ij, (3, 2)).place(x)    # 1 (default)
+    ti.root.dense(ti.ji, (2, 3)).place(y)    # 2
+
+They can be accessed in the same manner: ``x[i, j]`` and ``y[i, j]``.
```

Comment:
Logically speaking, both `x` and `y` have the same shape of `(3, 2)`, and they can be accessed ..., where `0 <= i < 3 && 0 <= j < 2`?